### PR TITLE
Add a check for non-pointer types in sync.Pool.Put

### PR DESCRIPTION
**Pull request by bradleyfalzon**
Implements #10

### CODE REVIEW

File: lint.go [hunk 1]
```diff
@@ -715,6 +716,37 @@ func CheckEarlyDefer(f *lint.File) {
...
+
+		sel, ok := call.Fun.(*ast.SelectorExpr)
+		if !ok {
+			return true
+		}
+		if sel.Sel.Name != "Put" {
+			return true
+		}
+		typ := f.Pkg.TypesInfo.TypeOf(sel.X)
+		if typ == nil || typ.String() != "sync.Pool" {
```
Comment (dominikh) [inline lint.go]: This might be a `*sync.Pool`, too.

File: lint.go [hunk 2]
```diff
@@ -715,6 +716,37 @@ func CheckEarlyDefer(f *lint.File) {
...
+		}
+
+		arg := f.Pkg.TypesInfo.TypeOf(call.Args[0]).Underlying()
+		switch arg.(type) {
+		case *types.Interface, *types.Pointer, *types.Map, *types.Chan:
+			// ignore interfaces as they're non-trivial to follow currently
+			return true
+		}
+		f.Errorf(call.Args[0], 1, "Non-pointer type %T put into sync.Pool", arg)
+		return true
```
Comment (dominikh) [inline lint.go]: You can return `false` here.

File: lint.go [hunk 3]
```diff
@@ -715,6 +716,37 @@ func CheckEarlyDefer(f *lint.File) {
...
+		}
+		typ := f.Pkg.TypesInfo.TypeOf(sel.X)
+		if typ == nil || typ.String() != "sync.Pool" {
+			return true
+		}
+
+		arg := f.Pkg.TypesInfo.TypeOf(call.Args[0]).Underlying()
+		switch arg.(type) {
+		case *types.Interface, *types.Pointer, *types.Map, *types.Chan:
+			// ignore interfaces as they're non-trivial to follow currently
```
Comment (dominikh) [inline lint.go]: This comment doesn't match what we're doing.

File: lint.go [hunk 4]
```diff
@@ -37,6 +37,7 @@ var Funcs = []lint.Func{
 	CheckURLs,
 	CheckEarlyDefer,
 	CheckEmptyCriticalSection,
+	CheckSyncPoolPointers,
```
Comment (dominikh) [inline lint.go]: We probably want to put this check in the list of dubious checks instead.

File: README.md [hunk 5]
```diff
@@ -91,6 +91,7 @@ The following things are currently checked by staticcheck:
 - Checks for empty critical sections, e.g. `(*sync.Mutex).Lock`
   directly followed by `(*sync.Mutex).Unlock`. This usually indicates
   a missing `defer`, or otherwise questionable code.
+- `sync.Pool.Put()` - Checks the argument is a pointer type.
```
Comment (dominikh) [inline README.md]: Rephrase as follows: `Checks that arguments to sync.Pool.Put are pointers`.

File: lint.go [hunk 6]
```diff
@@ -750,6 +751,37 @@ func CheckEarlyDefer(f *lint.File) {
...
+
+		sel, ok := call.Fun.(*ast.SelectorExpr)
+		if !ok {
+			return true
+		}
+		if sel.Sel.Name != "Put" {
+			return true
+		}
+		typ := f.Pkg.TypesInfo.TypeOf(sel.X)
+		if typ == nil || (typ.String() != "sync.Pool" && typ.String() != "*sync.Pool") {
```
Comment (bradleyfalzon) [inline lint.go:769]: This is my proposed solution to permit both `sync.Pool` and `*.sync.Pool`, there's other ways to do this, but I thought this was relatively readable and straight forward, but it's also not perfectly tidy, so happy to hear alternatives.

File: testdata/dubious/sync-pool-pointers.go [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package pkg
+
+import "sync"
+
+func fn() {
+	pool := sync.Pool{}
```
Comment (bradleyfalzon) [inline testdata/dubious/sync-pool-pointers.go]: I'm not checking for `*sync.Pool` here, do you want this test duplicated to include that?

**Comment by bradleyfalzon** [comment]
Updated with additional comments.

File: testdata/dubious/sync-pool-pointers.go (hunk 7 above)
Comment (dominikh) [inline testdata/dubious/sync-pool-pointers.go]: Yes please.

File: lint.go (hunk 6 above)
Comment (dominikh) [inline lint.go:769]: Looks fine, thanks.

**Comment by bradleyfalzon** [comment]
Updated with additional tests.

**Pull request closed without merging**